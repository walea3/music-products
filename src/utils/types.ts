export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};
