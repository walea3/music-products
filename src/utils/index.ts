export * from './product';
export * from './useAPI';
export * from './test';
