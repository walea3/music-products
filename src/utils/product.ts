export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};

export const BASE_PRICE = 12.99;
export const DEFAULT_BG_COLOR = '#424242';
