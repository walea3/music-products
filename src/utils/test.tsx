import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Route, RouteProps, BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export const renderPage = (
  page: RouteProps['children'],
  path: RouteProps['path'],
  route: string,
) => renderWithRouter(<Route path={path}>{page}</Route>, { route });

// based on react-testing library default
const TEST_ID_ATTR = 'data-testid';
const TEST_ID_DELIMITER = '-';

// @TODO if desired remove test id in prod
export const testIdProp = (baseTestId: string) => (id?: string) => ({
  [TEST_ID_ATTR]: `${baseTestId}${id ? `${TEST_ID_DELIMITER}${id}` : ''}`,
});

export const getTestId = (...args: any[]) =>
  args.filter(Boolean).join(TEST_ID_DELIMITER);
