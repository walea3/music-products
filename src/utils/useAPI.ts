import { useState, useEffect, useCallback } from 'react';

export const useAPI = <T>(url: string) => {
  const [error, setError] = useState(null);
  const [data, setData] = useState<T>();
  const [isLoading, setLoading] = useState(true);

  const fetchData = useCallback(async () => {
    setLoading(true);
    setError(null);

    try {
      const res = await fetch(`/api/${url}`);
      const data: T = await res.json();

      setData(data);
    } catch (error) {
      setError(error);
    } finally {
      setLoading(false);
    }
  }, [url]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return [isLoading, data, error, fetchData] as const;
};
