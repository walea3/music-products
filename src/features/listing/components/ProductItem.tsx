import { ReactNode } from 'react';
import styled from 'styled-components';

import { BASE_PRICE, DEFAULT_BG_COLOR, Product, testIdProp } from 'utils';

export function ProductItem({
  id,
  title,
  children,
  artist,
  price,
  image1,
  textColor,
  accentColor,
  backgroundColor = DEFAULT_BG_COLOR,
  isFree,
}: Product & { children?: ReactNode }) {
  const testId = testIdProp(`Product-${id}`);

  return (
    <ProductContainer
      $color={textColor}
      $borderColor={accentColor}
      {...testId()}
    >
      <ProductImage src={image1} alt={artist} {...testId('image')} />
      <ProductInfo $color={backgroundColor}>
        <ProductTitle {...testId('title')}>{title}</ProductTitle>
        {children}
        <ProductPrice $color={accentColor} {...testId('price')}>
          {isFree ? 'Free' : <>&pound;{price || BASE_PRICE}</>}
        </ProductPrice>
      </ProductInfo>
    </ProductContainer>
  );
}

const ProductImage = styled.img`
  display: block;
  width: 100%;
  height: auto;
  margin: 0;
  transition: transform 0.5s ease-in-out;
`;

const ProductContainer = styled.article<{
  $color: Product['textColor'];
  $borderColor: Product['accentColor'];
}>`
  position: relative;
  display: inline-block;
  flex-direction: column;
  border-top: 5px solid;
  border-color: ${(props) => props.$borderColor};
  overflow: hidden;
  width: 90%;
  max-width: 490px;
  color: ${(props) => props.$color};

  &:hover ${ProductImage} {
    transform: scale(1.2);
  }
`;

const ProductInfo = styled.div<{
  $color: Product['backgroundColor'];
}>`
  position: relative;
  z-index: 1;
  padding: 20px;
  background-color: ${(props) => props.$color};
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
`;

const ProductTitle = styled.h2`
  margin-top: 0;
`;

const ProductPrice = styled.div<{
  $color: Product['accentColor'];
}>`
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: ${(props) => props.$color};
  font-weight: bold;
  font-size: 1.2rem;
`;
