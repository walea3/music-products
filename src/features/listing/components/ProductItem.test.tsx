import React from 'react';
import { render, RenderResult } from '@testing-library/react';

import { getTestId } from 'utils';

import { ProductItem } from './ProductItem';

describe('<ProductItem />', () => {
  const requiredProps = {
    id: 123,
    title: 'Product title',
    artist: 'Artist name',
    description: 'Description',
    isFree: true,
    image1: 'path/to/image1.jpg',
    image2: 'path/to/image2.jpg',
    accentColor: '#00b6b8',
    backgroundColor: '#F2F2F2',
    textColor: '#08070b',
    feature_order: '0',
  };

  const findByTestId = (
    wrapper: RenderResult,
    { id }: typeof requiredProps,
    childTestId?: string,
  ) => wrapper.getByTestId(getTestId('Product', id, childTestId));

  describe('@renders', () => {
    test('a free product as default', () => {
      const wrapper = render(<ProductItem {...requiredProps} />);

      expect(findByTestId(wrapper, requiredProps)).toMatchSnapshot();
    });

    test('a paid product without a price', () => {
      const wrapper = render(<ProductItem {...requiredProps} isFree={false} />);

      expect(findByTestId(wrapper, requiredProps, 'price')).toMatchSnapshot();
    });

    test('a paid product with a price', () => {
      const wrapper = render(
        <ProductItem {...requiredProps} isFree={false} price="9.99" />,
      );

      expect(findByTestId(wrapper, requiredProps, 'price')).toMatchSnapshot();
    });

    test('without background color', () => {
      const wrapper = render(
        <ProductItem {...requiredProps} isFree={false} price="9.99" />,
      );

      expect(findByTestId(wrapper, requiredProps, 'price')).toMatchSnapshot();
    });
  });
});
