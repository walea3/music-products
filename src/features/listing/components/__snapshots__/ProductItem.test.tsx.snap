// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> @renders a free product as default 1`] = `
.c2 {
  display: block;
  width: 100%;
  height: auto;
  margin: 0;
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c0 {
  position: relative;
  display: inline-block;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top: 5px solid;
  border-color: #00b6b8;
  overflow: hidden;
  width: 90%;
  max-width: 490px;
  color: #08070b;
}

.c0:hover .c1 {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c3 {
  position: relative;
  z-index: 1;
  padding: 20px;
  background-color: #F2F2F2;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
}

.c4 {
  margin-top: 0;
}

.c5 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

<article
  class="c0"
  data-testid="Product-123"
>
  <img
    alt="Artist name"
    class="c1 c2"
    data-testid="Product-123-image"
    src="path/to/image1.jpg"
  />
  <div
    class="c3"
  >
    <h2
      class="c4"
      data-testid="Product-123-title"
    >
      Product title
    </h2>
    <div
      class="c5"
      data-testid="Product-123-price"
    >
      Free
    </div>
  </div>
</article>
`;

exports[`<ProductItem /> @renders a paid product with a price 1`] = `
.c0 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

<div
  class="c0"
  data-testid="Product-123-price"
>
  £
  9.99
</div>
`;

exports[`<ProductItem /> @renders a paid product without a price 1`] = `
.c0 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

<div
  class="c0"
  data-testid="Product-123-price"
>
  £
  12.99
</div>
`;

exports[`<ProductItem /> @renders without background color 1`] = `
.c0 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

<div
  class="c0"
  data-testid="Product-123-price"
>
  £
  9.99
</div>
`;
